metriken-exposition/src/parquet.rs

```diff
@@ -0,0 +1,277 @@
+use std::collections::BTreeMap;
+use std::fs::File;
+use std::sync::Arc;
+
+use crate::snapshot::{HashedSnapshot, Snapshot};
+use crate::HistogramSnapshot;
+use arrow::array::*;
+use arrow::datatypes::*;
+use arrow::error::ArrowError;
+use parquet::arrow::ArrowWriter;
+use parquet::basic::{Compression, ZstdLevel};
+use parquet::errors::ParquetError;
+use parquet::file::properties::WriterProperties;
+use parquet::format::FileMetaData;
+
+pub struct ParquetWriter {
+    /// File handle for output parquet file
+    file: Option<File>,
+    writer: Option<ArrowWriter<File>>,
+    properties: WriterProperties,
+
+    /// Number of rows cached in memory before being written as a `RecordBatch`
+    max_batch_size: usize,
+
+    /// Schema and columnar data for timestamps
+    tses: Vec<u64>,
+
+    /// Schema and columnar data for counters
+    counters: BTreeMap<String, Vec<Option<u64>>>,
+
+    /// Schema and columnar data for gauges
+    gauges: BTreeMap<String, Vec<Option<i64>>>,
+
+    /// Schema and columnar data for histograms
+    histograms: BTreeMap<String, Vec<Option<HistogramSnapshot>>>,
+
+    /// Schema generated for the parquet file after finalizing metadata
+    schema: Arc<Schema>,
+}
+
+/// Convert metrics snapshot data into parquet
+///
+/// Parquet files need the entire schema to be known prior to starting to
+/// write any data and all the columns to be of the same length. Missing
+/// data can be filled-up using NULL/empty values (columns can be marked
+/// nullable), but must be accounted for.
+///
+/// Metrics snapshot data does not have a well-defined schema, i.e., dynamic
+/// metrics may appear in some snapshots, but not in others, all the snapshots
+/// must be scanned and a complete list of all metrics extracted before
+/// the parquet file can be written. Further, since the parquet file may be
+/// too large to fit into memory, snapshots should be processed in batches.
+///
+/// This means that we need to make two passes over the snapshots:
+/// a) First, we collect the cumulative union of all the metrics that have
+///    been seen in any snapshot and build the schema.
+/// b) We process batches of snapshots extracing the row-centric data and
+///    converting it to a columnar form while filling in `None` values for
+///    metrics which are missing in specific snapshots.
+impl ParquetWriter {
+    const DEFAULT_MAX_BATCH_SIZE: usize = 1024 * 1024;
+
+    pub fn try_new(file: File, compression: bool) -> Result<Self, ParquetError> {
```

Comment:
I'm not sure I like this interface. I think that should probably have a signature that's closer to std::fs::File::open() 

```
pub fn open<P: AsRef<Path>>(path: P) -> Result<File> { ... }
```

You might also consider using the builder pattern here so that we don't need try_new_... for each additional optional parameter.